src/lib.rs

```diff
@@ -104,86 +105,127 @@ mod parser;
 const VERSION: &str = "0.0.1";
 const LANG: &str = "rust";
 
-impl Options {
-    /// Add a name option for the unconnected connection.
+#[doc(hidden)]
+pub mod options_typestate {
+    /// `Options` typestate indicating
+    /// that there has not yet been
+    /// any auth-related configuration
+    /// provided yet.
+    #[derive(Debug, Copy, Clone, Default)]
+    pub struct Unauthenticated;
+
+    /// `Options` typestate indicating
+    /// that auth-related configuration
+    /// has been provided, and may not
+    /// be provided again.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Authenticated;
+
+    /// `Options` typestate indicating that
+    /// this `Options` has been used to create
+    /// a `Connection` and may not be changed.
+    #[derive(Debug, Copy, Clone)]
+    pub struct Finalized;
+}
+
+/// A configuration object for a NATS connection.
+#[derive(Clone, Debug, Default)]
+pub struct Options<Typestate = options_typestate::Unauthenticated> {
+    auth: AuthStyle,
+    name: Option<String>,
+    no_echo: bool,
```

Comment:
FYI we have lots of these, the ones here were just the quick ones I chose to test out first.